tests/test_rng_state.py

```diff
@@ -49,12 +50,64 @@ def run_rng_state(Model):
     assert state_1a == state_1
 
 
+def run_rng_state_with_bitgen(Model, bitgen_class, seed):
+    """Test RNG state save/restore with specific bit generator."""
+    if Model is im.Sophia20:
+        evt_kin = FixedTarget(13 * TeV, "photon", "proton")
+    elif Model is im.UrQMD34:
+        evt_kin = CenterOfMass(50 * GeV, "proton", "proton")
+    else:
+        evt_kin = CenterOfMass(13 * TeV, "proton", "proton")
+
+    rng = np.random.Generator(bitgen_class(seed=seed))
+    generator = Model(evt_kin, seed=rng)
+
+    nevents = 10 if Model is not im.EposLHC else 50
+
+    state_0 = deepcopy(generator.random_state)
+    states = [deepcopy(generator.random_state) for _ in generator(nevents)]
+
+    generator.random_state = state_0
+
+    for i, _ in enumerate(generator(nevents)):
+        current_state = generator.random_state
+        # Use str() for comparison to handle numpy arrays in nested dicts
+        assert str(states[i]) == str(
+            current_state
+        ), f"states differ after {i+1} generation with {bitgen_class.__name__}"
+
+
 @pytest.mark.parametrize("Model", get_all_models())
 def test_rng_state(Model):
-    if Model is im.Pythia8:
-        pytest.skip("Pythia8 currently does not support rng_state serialization")
-
     if Model in (im.UrQMD34,):
         pytest.xfail(f"{Model.pyname} fails this test, needs investigation")
+    # Skip EposLHCR models - they have pre-existing Fortran bugs unrelated to RNG
+    if Model in (im.EposLHCR, im.EposLHCRHadrRescattering):
+        pytest.skip(f"{Model.pyname} has known Fortran issues (negative ked)")
 
     run_in_separate_process(run_rng_state, Model)
+
+
+@pytest.mark.parametrize("Model", get_all_models())
+@pytest.mark.parametrize(
+    "bitgen_class,seed",
+    [
+        (np.random.PCG64, 11111),
+        (np.random.MT19937, 22222),
+        (np.random.Philox, 33333),
+        (np.random.SFC64, 44444),
+        (np.random.PCG64DXSM, 55555),
+    ],
+)
+def test_rng_state_bitgens(Model, bitgen_class, seed):
+    """Test different NumPy bit generators with all models."""
+    if Model in (im.UrQMD34,):
+        pytest.xfail(f"{Model.pyname} fails this test, needs investigation")
+    # Skip EposLHCR models - they have pre-existing Fortran bugs unrelated to RNG
+    if Model in (im.EposLHCR, im.EposLHCRHadrRescattering):
+        pytest.skip(f"{Model.pyname} has known Fortran issues (negative ked)")
```

Comment:
what does "negative ked" mean?

Reply:
It is an old message. "ked" is net down quarks (d quarks - d antiquarks), but the problem doesn't appear now. The message now provides the real problem related to usage of UrQMD.